File: call_bridge/identity.py
"""LiveKit identity hash functions.

These must produce identical output to lk-jwt-service's Go implementation.
The separator is a literal pipe character ``|``.  Output is unpadded base64.
"""

import base64
import hashlib


def _hash_unpadded_b64(data: str) -> str:
    """SHA-256 hash *data* (UTF-8) and return unpadded base64."""
    digest = hashlib.sha256(data.encode()).digest()
    return base64.b64encode(digest).decode().rstrip("=")


def livekit_room_alias(matrix_room_id: str) -> str:
    """Derive the LiveKit room name from a Matrix room ID.

    Formula: ``base64_unpadded(SHA256(roomId | "m.call#ROOM"))``

    The ``|`` is a literal pipe separator, matching lk-jwt-service and MSC4195.
    """
    return _hash_unpadded_b64(f"{matrix_room_id}|m.call#ROOM")


def livekit_identity(user_id: str, device_id: str, session_id: str) -> str:
    """Derive a LiveKit participant identity from a membership triple.

    Formula: ``base64_unpadded(SHA256(userId | deviceId | sessionId))``
    """
    return _hash_unpadded_b64(f"{user_id}|{device_id}|{session_id}")


def livekit_breakout_alias(matrix_room_id: str, breakout_id: str) -> str:
    """Derive a LiveKit room name for a breakout room.

    Formula: ``base64_unpadded(SHA256(roomId | "m.call#BREAKOUT#" + breakoutId))``
    """
    return _hash_unpadded_b64(f"{matrix_room_id}|m.call#BREAKOUT#{breakout_id}")

File: identity.go
package main

import (
	"crypto/sha256"
	"encoding/base64"
)

// hashUnpaddedBase64 computes SHA-256 of data and returns unpadded standard base64.
func hashUnpaddedBase64(data string) string {
	h := sha256.Sum256([]byte(data))
	return base64.RawStdEncoding.EncodeToString(h[:])
}

// LiveKitRoomAlias derives the LiveKit room name from a Matrix room ID
// using the MSC4195 hash formula: base64_unpadded(SHA256(roomId | "m.call#ROOM")).
func LiveKitRoomAlias(matrixRoomID string) string {
	return hashUnpaddedBase64(matrixRoomID + "|m.call#ROOM")
}

// LiveKitRoomAliasForMode returns the LiveKit room name based on the alias mode.
// "raw" = use Matrix room ID directly (older lk-jwt-service).
// "hash" = SHA256 hash per MSC4195 (newer lk-jwt-service).
func LiveKitRoomAliasForMode(matrixRoomID, mode string) string {
	if mode == "hash" {
		return LiveKitRoomAlias(matrixRoomID)
	}
	return matrixRoomID // raw mode — pass through
}

// LiveKitIdentity returns the LiveKit participant identity for a call member.
//
// Element Call expects the LK identity to match the format it reads from
// call.member state events: "userId:deviceId". If the identity doesn't match,
// EC's MatrixAudioRenderer refuses to render the participant's tracks.
func LiveKitIdentity(userID, deviceID string) string {
	return userID + ":" + deviceID
}

// LiveKitBreakoutAlias derives a LiveKit room name for a breakout room.
//
// Formula: base64_unpadded(SHA256(roomId | "m.call#BREAKOUT#" + breakoutId))
func LiveKitBreakoutAlias(matrixRoomID, breakoutID string) string {
	return hashUnpaddedBase64(matrixRoomID + "|m.call#BREAKOUT#" + breakoutID)
}
